arc.mm

```diff
@@ -838,8 +838,9 @@ static BOOL setObjectHasWeakRefs(id obj)
 	WeakRef *oldRef;
 	id old;
 	loadWeakPointer(addr, &old, &oldRef);
-	// If the old and new values are the same, then we don't need to do anything.
-	if (old == obj)
+	// If the old and new values are the same, then we don't need to do anything
+	// unless we are deleting the weak reference by storing NULL to it.
+	if (old == obj && (obj != NULL || NULL == oldRef))
```

Comment:
Please can you add some brackets around this? I am not sure what the precedenc e is between || and == without looking it up and the next person reading the code might not either.

Reply:
Done.